### PR TITLE
Mostly read-only REST API

### PR DESCRIPTION
**Pull request by bonzini**
This pull request implements the beginning of #25 ("Add REST API").
The API is minimal on the write side (no login, no posting of messages), but it already includes:
* the possibility to search and browse the database
* a complex schema to easily retrieve the metadata for entire threads (aka "series") with a limited number of API calls
* plugin integration
* an embryo of a first-class "results" model, used by the git and testing plugins to hide the internal properties-based implementation

### CODE REVIEW

**Comment by bonzini** [comment]
The failure is only in Python 3.7 nightly builds, and is due to Django, not this pull request:

      File "/home/travis/virtualenv/python3.7-dev/lib/python3.7/site-packages/django/contrib/admin/widgets.py", line 152
        '%s=%s' % (k, v) for k, v in params.items(),
        ^
    SyntaxError: Generator expression must be parenthesized

File: api/rest.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    """
+    Allows access only to admin users.
+    """
+    def has_permission(self, request, view):
+        return request.method in permissions.SAFE_METHODS or \
+               (request.user and request.user.is_superuser)
+
+class IsMaintainerOrReadOnly(permissions.BasePermission):
+    """
+    Allows access only to admin users.
```
Comment (famz) [inline api/rest.py]: "only to maintainers"?

File: api/rest.py [hunk 2]
```diff
@@ -127,11 +153,26 @@ class SeriesSerializer(BaseMessageSerializer):
...
     resource_uri = HyperlinkedMessageField(view_name='series-detail')
     message = HyperlinkedMessageField(view_name='messages-detail')
 
+    def get_results(self, message):
+        results = {}
+        dispatch_module_hook("rest_results_hook", message=message, results=results)
+        return results
+    results = SerializerMethodField()
+
+    def __init__(self, *args, **kwargs):
```
Comment (famz) [inline api/rest.py]: Is it easier to read if we move __init__ to the first method in the class?

**Comment by famz** [comment]
I haven't looked in depth into django-rest-framework and the nested router plugin. But the PR looks great in general.

The question I have is testing. Is the plan to swtich patchew-cli or the web interface to be a client? (I think both are good things) And are you going to add cases to tests/?

**Comment by bonzini** [comment]
> Is the plan to swtich patchew-cli or the web interface to be a client? (I  think both are good things)

Yes and no respectively (I prefer the web interface to do everything server-side, I don't like single-page web apps and they are beyond my limited knowledge of Django!).

However, even for patchew-cli, people are using patchew-project/patchew's patchew-cli as a canonical source of "a command line tha works with patchew.org" so that should only be done after the REST API is deployed on patchew.org.

> And are you going to add cases to tests/?

Yes, and I'll do that before pushing this.

File: api/rest.py (hunk 1 above)
Comment (bonzini) [reply api/rest.py]: "only to admin users or maintainers"

**Comment by bonzini** [comment]
Needless to say, adding tests found some bugs. :) Comparison between v1 and v2:

https://github.com/bonzini/patchew/compare/rest-v1...bonzini:rest?expand=1

File: tests/test_rest.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
+#!/usr/bin/env python3
+#
+# Copyright 2017 Red Hat, Inc.
+#
+# Authors:
+#     Fam Zheng <famz@redhat.com>
```
Comment (famz) [inline tests/test_rest.py]: I don't mind but you can update the year and the name if you want. :)

File: tests/test_git.py [hunk 4]
```diff
@@ -100,6 +102,51 @@ def test_apply_with_base_and_brackets(self):
...
+Applying: bar patch
+error: sha1 information is lacking or useless (foo).
+error: could not build fake ancestor
+Patch failed at 0001 bar patch
+The copy of the patch that failed is found in: .git/rebase-apply/patch
+When you have resolved this problem, run "git am --continue".
+If you prefer to skip this patch, run "git am --skip" instead.
+To restore the original branch and stop patching, run "git am --abort".
+Failed to apply patch:
+[Qemu-devel] [PATCH] bar patch""")
```
Comment (famz) [inline tests/test_git.py]: I think the actual error message depends on the git version. I'm getting this on Fedora (git version 2.14.3):

```
Switched to a new branch '20160628014747.20971-2-famz@redhat.com'
Branch 20160628014747.20971-2-famz@redhat.com set up to track local branch test by rebasing.
Applying: bar patch
error: sha1 information is lacking or useless (foo).
error: could not build fake ancestor
Patch failed at 0001 bar patch
The copy of the patch that failed is found in: .git/rebase-apply/patch
When you have resolved this problem, run "git am --continue".
If you prefer to skip this patch, run "git am --skip" instead.
To restore the original branch and stop patching, run "git am --abort".
Failed to apply patch:
[Qemu-devel] [PATCH] bar patch
```

I.e. there is one extra line in the middle:

```
Branch 20160628014747.20971-2-famz@redhat.com set up to track local branch test by rebasing.
```
Maybe simply remove the `self.assertEquals(log.content, expected_log)` check?

File: tests/test_git.py [hunk 5]
```diff
@@ -100,6 +102,51 @@ def test_apply_with_base_and_brackets(self):
...
+        self.assertEqual(resp.data['results'][0]['results']['git']['status'], 'success')
+        self.assertEqual(resp.data['results'][0]['results']['git']['repo'], self.repo)
+        self.assertEqual(resp.data['results'][0]['results']['git']['tag'], "refs/tags/patchew/20160628014747.20971-1-famz@redhat.com")
+        log = self.client.get(resp.data['results'][0]['results']['git']['log_url'])
+        expected_log = self.git_log("""Switched to a new branch '20160628014747.20971-1-famz@redhat.com'
+Applying: quorum: Only compile when supported
+To XXXXXX/repo
+ * [new tag]         patchew/20160628014747.20971-1-famz@redhat.com -> patchew/20160628014747.20971-1-famz@redhat.com
+""")
+        self.assertEquals(log.content, expected_log)
```
Comment (famz) [inline tests/test_git.py]: The same error here for me.

**Comment by famz** [comment]
I am done with an incremental review, the code looks good. Thank you for working on this, and adding tests! I think we are almost ready to go!

File: tests/test_git.py (hunk 4 above)
Comment (bonzini) [reply tests/test_git.py]: I'll change it to a status code test (must be 200).

**Comment by bonzini** [comment]
Done! As before, for an incremental review you can use branch `rest-v2` (I rebased so that travis passes, but the changes to the series are identified easily).

As discussed on IRC, I've added a migration to fix UTF-8 recipients and a test case for that too. You still cannot search for UTF-8 recipients because of how they are encoded in JSON works, but that's a separate bug in `api.search`.

**Comment by famz** [comment]
Looks good. Except this error when migrating the live patchew.org database (let me know if you want access the server, or alternatively):

```
(venv) fam@lemon:~/work/patchew$ ./manage.py migrate
Loaded module: footer
Loaded module: diff
Loaded module: email
Loaded module: git
Loaded module: tags
Loaded module: testing
Operations to perform:
  Apply all migrations: admin, api, auth, contenttypes, sessions
Running migrations:
  Applying api.0020_auto_20180204_0647... OK
  Applying api.0021_fix_recipients_utf8...Traceback (most recent call last):
  File "./manage.py", line 19, in <module>
    execute_from_command_line(sys.argv)
  File "/stor/work/patchew/venv/lib64/python3.6/site-packages/django/core/management/__init__.py", line 364, in execute_from_command_line
    utility.execute()
  File "/stor/work/patchew/venv/lib64/python3.6/site-packages/django/core/management/__init__.py", line 356, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/stor/work/patchew/venv/lib64/python3.6/site-packages/django/core/management/base.py", line 283, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/stor/work/patchew/venv/lib64/python3.6/site-packages/django/core/management/base.py", line 330, in execute
    output = self.handle(*args, **options)
  File "/stor/work/patchew/venv/lib64/python3.6/site-packages/django/core/management/commands/migrate.py", line 204, in handle
    fake_initial=fake_initial,
  File "/stor/work/patchew/venv/lib64/python3.6/site-packages/django/db/migrations/executor.py", line 115, in migrate
    state = self._migrate_all_forwards(state, plan, full_plan, fake=fake, fake_initial=fake_initial)
  File "/stor/work/patchew/venv/lib64/python3.6/site-packages/django/db/migrations/executor.py", line 145, in _migrate_all_forwards
    state = self.apply_migration(state, migration, fake=fake, fake_initial=fake_initial)
  File "/stor/work/patchew/venv/lib64/python3.6/site-packages/django/db/migrations/executor.py", line 244, in apply_migration
    state = migration.apply(state, schema_editor)
  File "/stor/work/patchew/venv/lib64/python3.6/site-packages/django/db/migrations/migration.py", line 129, in apply
    operation.database_forwards(self.app_label, schema_editor, old_state, project_state)
  File "/stor/work/patchew/venv/lib64/python3.6/site-packages/django/db/migrations/operations/special.py", line 193, in database_forwards
    self.code(from_state.apps, schema_editor)
  File "/stor/work/patchew/api/migrations/0021_fix_recipients_utf8.py", line 24, in fix_utf8_recipients
    recipients = [[_parse_header(x[0]), x[1]] for x in recipients]
  File "/stor/work/patchew/api/migrations/0021_fix_recipients_utf8.py", line 24, in <listcomp>
    recipients = [[_parse_header(x[0]), x[1]] for x in recipients]
  File "/stor/work/patchew/api/migrations/0021_fix_recipients_utf8.py", line 12, in _parse_header
    h = h.decode(c or 'utf-8')
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xed in position 8: invalid continuation byte
```

**Comment by bonzini** [comment]
I'll drop the last patch then. I think the culprit is [this patch](http://patchew.org/QEMU/20170615220501.GA26408@flamenco/mbox).

**Comment by bonzini** [comment]
Oh, it's actually a wrong record in the database:

                    "name": "=?UTF-8?q?Daniel=20D=EDaz?=",
                    "address": "daniel.diaz@linaro.org"

Thunderbird "correctly" shows it as _Daniel D�az_.

**Pull request closed without merging**